### ISSUE TITLE
Producer message delivery fails when broker connectivity is down

### ISSUE THREAD

**Issue by krishanfrog** [edenhill/librdkafka#1794]
How to reproduce
===========
I have a cluster of 4 brokers with a replication factor of 2 on all topics.

One of the broker loses network connectivity with all external services. The process is up but the instance just loses network connectivity for 10 minutes.

The group coordinator recognizes this and reassigns the leaders of the partitions assigned to this broker.

But none of the producer messages for these partitions are delivered. All messages throw an error

```
required feature not supported by broker
```

Once the connectivity is back up the leaders are reassigned back and now new messages are delivered.

Why did the messages fail to be sent to the new leader during the issue window? Is the flow different depending if the broker is down versus broker is up but not able to connect

Checklist
=========

Producer librdkafka version - 0.11.3
Apache Kafka version: 1.0.1

```
producer configuration: {
		"metadata.broker.list": "X:9092,Y:9092,Z:9092,U:9092",
		"max.in.flight.requests.per.connection": 5,
		"queue.buffering.max.messages": 50000,
		"queue.buffering.max.kbytes": 153600,
		"queue.buffering.max.ms": 100,
		"message.send.max.retries": 5,
		"retry.backoff.ms": 100,
		"batch.num.messages": 50,
		"dr_msg_cb": true,
		"log.connection.close": false,
		"socket.keepalive.enable": true,
		"statistics.interval.ms": 60000
	}
"topicConfig": {
		"request.required.acks": 1,
		"message.timeout.ms": 900000
	}
```
Operating system: Ubuntu 14.04

**Comment by krishanfrog** [comment]
Some more information 

I have 40 producers. The produce message rate is not completely uniform across all producers. It ranges from 1000 messages per sec to 2500 messages per sec in these producers.

Issue 1 - Messages dropped
For 10 of these producers all messages attempted during the 1st 3 minutes of the issue window dropped and were not delivered (I got a failed delivery report for them after 3 minutes). After 3 minutes new messages were delivered to the new leader in the kafka cluster.

Issue 2 - Messages delivered out of order
For the other 30 producers no messages were delivered during the 1st 3 minutes of the issue window. After 3 minutes new messages were delivered to the new leader in the kafka cluster. After the original broker was back up the old messages in the 1st 3 minute window are now delivered 

@edenhill What are the reasons for these behaviours?

Is it expected to take upto 3 minutes to direct messages to the new leader of the cluster? I have verified that the leader rebalance happened immediately after the broker lost connectivity

**Comment by krishanfrog** [comment]
The out of order issue is not happening with V0.11.4

But I am worried about why certain messages were dropped after 3 minutes even though message.timeout.ms is set to 15 minutes and why it took upto 3 minutes to direct messages to the new leader of the partitions?

**Comment by edenhill** [comment]
We did a bunch of producer retry fixes for v0.11.4 and also documented the behaviour.

The premature timeout could be because the retries are exhausted.

https://github.com/edenhill/librdkafka/blob/master/INTRODUCTION.md#producer-message-delivery-failure

**Comment by krishanfrog** [comment]
But shouldn't the retries be tried over the time range starting from retry.backoff.ms to message.timeout.ms? 

I believe the message.send.max.retries number is distributed across this time range.

Is this not how retry works? If not is the retry behaviour (count and time backoff) documented somewhere? Want to understand so that can set the right values for these parameters

Also Why did it take upto 3 minutes to direct messages to the new leader of the partitions?

**Comment by edenhill** [comment]
`message.timeout.ms` depicts the maximum time librdkafka may use to deliver a message, the retry handling happens within this period and the message fails permanently when either the retry count or message timeout is exceeded.

Retries are triggered after retry.backoff.ms, but only if the leader broker for the given partition is available, otherwise librdkafka falls back to polling the topic metadata to monitor when a new leader is elected and then perform a retry. The metadata polling is controlled first by `topic.metadata.refresh.fast.interval.ms` followed by `topic.metadata.refresh.interval.ms`.
https://github.com/edenhill/librdkafka/blob/master/CONFIGURATION.md

**Comment by krishanfrog** [comment]
Got it. Can you describe the error message in the delivery failure callback. Why would I get this particular error message. Couldn't find this in the documentation.

```
required feature not supported by broker
```

Also a suggestion - 
I see that topic.metadata.refresh.fast.interval.ms is exponentially backed off until it reaches topic.metadata.refresh.interval.ms. I suggest to keep another attribute for the exponential backoff limit. Because I would want to keep a high topic.metadata.refresh.interval.ms to not increase the load on the brokers, but would also want to keep polling for metadata when a broker goes down but this polling interval should not exceed a certain time limit so that things can move on.
Eg:
topic.metadata.refresh.interval.ms - 5 minutes
topic.metadata.refresh.fast.interval.ms - 250 ms
topic.metadata.refresh.fast.interval.max.ms - 30 seconds

Let me know if this makes sense and I can raise a PR for the same too.

**Comment by DuncanSands** [comment]
"required feature not supported by broker" looks like issue 1694 which was fixed in release 0.11.4.

**Comment by edenhill** [comment]
@krishanfrog You have `topic.metadata.refresh.fast.cnt` for this purpose
